{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr  source        type  start    end score strand phase  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "             gene_id                           gene_type gene_name level  \n",
       "0  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2  \n",
       "1  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2  \n",
       "2  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2  \n",
       "3  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2  \n",
       "4  ENSG00000223972.5  transcribed_unprocessed_pseudogene   DDX11L1     2  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the data\n",
    "df = pd.read_csv('data/gencode.v28.annotation.gtf', sep='\\t', comment='#', header=None)\n",
    "df.columns = ['chr', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "\n",
    "# Extract key-value pairs from the 'attributes' column\n",
    "attributes_df = df['attributes'].str.extract(r'gene_id \"(.*?)\";.*?gene_type \"(.*?)\";.*?gene_name \"(.*?)\";.*?level (\\d+);')\n",
    "attributes_df.columns = ['gene_id', 'gene_type', 'gene_name', 'level']\n",
    "\n",
    "# Concatenate the original DataFrame with the new attributes DataFrame\n",
    "df = pd.concat([df.iloc[:, :-1], attributes_df], axis=1)\n",
    "\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0m/90sjqm9d78l50t2x2lmdrztm0000gn/T/ipykernel_28069/929654069.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcripts['start'] = transcripts['start'] - 1500\n",
      "/var/folders/0m/90sjqm9d78l50t2x2lmdrztm0000gn/T/ipykernel_28069/929654069.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcripts['end'] = transcripts['start'] + 1500\n",
      "/var/folders/0m/90sjqm9d78l50t2x2lmdrztm0000gn/T/ipykernel_28069/929654069.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  transcripts['start'] = transcripts['start'].apply(lambda x: max(1, x))\n"
     ]
    }
   ],
   "source": [
    "transcripts = df[df['type'] == 'transcript']\n",
    "transcripts['start'] = transcripts['start'] - 1500\n",
    "transcripts['end'] = transcripts['start'] + 1500\n",
    "\n",
    "# ensure start positions are not less than 1\n",
    "transcripts['start'] = transcripts['start'].apply(lambda x: max(1, x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly select half of the promoters\n",
    "selected_promoters = transcripts.sample(frac=0.5)\n",
    "\n",
    "# Set reads to 0 for the other half\n",
    "transcripts.loc[~transcripts.index.isin(selected_promoters.index), 'reads'] = 0\n",
    "\n",
    "# Load the genome sequence\n",
    "genome = SeqIO.to_dict(SeqIO.parse('data/genome.fa', 'fasta'))\n",
    "\n",
    "counter = 0\n",
    "\n",
    "os.makedirs(\"output/no_reads\", exist_ok=True)\n",
    "os.makedirs(\"output/with_reads\", exist_ok=True)\n",
    "\n",
    "for index, row in transcripts.iterrows():\n",
    "    # Generate artificial 'reads' peak\n",
    "    if index in selected_promoters.index:\n",
    "        reads = np.random.poisson(10, (row['end'] - row['start'] + 1))\n",
    "        descriptor = 'with_reads'\n",
    "    else:\n",
    "        reads = np.zeros((row['end'] - row['start'] + 1))\n",
    "        descriptor = 'no_reads'\n",
    "    \n",
    "    # Extract genome sequence for the promoter region\n",
    "    chrom = row['chr']\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "    sequence = genome[chrom][start-1:end]\n",
    "    \n",
    "    # Create a DataFrame for the promoter region\n",
    "    promoter_df = pd.DataFrame({\n",
    "        'chrom': chrom,\n",
    "        'pos': np.arange(start, end+1),\n",
    "        'base': list(sequence),\n",
    "        'reads': reads,\n",
    "    })\n",
    "    \n",
    "    # Save the promoter DataFrame to a file\n",
    "    file_name = f'output/{descriptor}/promoter_{index}.feather'\n",
    "    promoter_df.to_feather(file_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
